### ISSUE TITLE
Improve S2259: Respect ! (null-forgiving) operator

### ISSUE THREAD

**Issue by martin-strecker-sonarsource** [SonarSource/sonar-dotnet#6082]
The [bang operator](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving) can be used to declare an expression as being not null. S2259 should not raise for expressions annotated with this operator.

```cs
#nullable enable
var o = null;
o!.ToString(); // Do not raise S2259 here
```

**Comment by pavel-mikula-sonarsource** [comment]
We should learn `NotNull` for such operation. And propagate it to tracked symbol, if exists.
